protocol/directory_test.go

```diff
@@ -138,3 +139,231 @@ func TestRegisterExistedUserWithoutTB(t *testing.T) {
+	if err != ErrorNameNotFound ||
+		res.Error != ErrorNameNotFound {
+		t.Fatal("Expect error code", ErrorNameNotFound, "got", err)
+	}
+	if ap := res.AP; ap == nil || bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		t.Fatal("Expect a proof of absence")
+	}
+	if res.TB != nil {
+		t.Fatal("Expect returned TB is nil")
+	}
+}
+
+func TestKeyLookupWithoutTB(t *testing.T) {
+	d := newDirectory(t, false)
+	_, err := d.Register(&RegistrationRequest{
+		Username: "alice",
+		Key:      []byte("key")})
+	if err != Success {
+		t.Fatal("Unable to register")
+	}
+	// lookup in the same epoch
+	// expect a proof of absence
+	// and error ErrorNameNotFound
+	res, err := d.KeyLookup(&KeyLookupRequest{Username: "alice"})
+	if err != ErrorNameNotFound ||
+		res.Error != ErrorNameNotFound {
+		t.Fatal("Expect error code", ErrorNameNotFound, "got", err)
+	}
+	if ap := res.AP; ap == nil || bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		t.Fatal("Expect a proof of absence")
+	}
+	if res.TB != nil {
+		t.Fatal("Expect returned TB is nil")
+	}
+
+	d.Update()
+	// lookup in epoch after registering epoch
+	// expect a proof of inclusion
+	res, err = d.KeyLookup(&KeyLookupRequest{Username: "alice"})
+	if ap := res.AP; ap == nil || !bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		t.Fatal("Expect a proof of inclusion")
+	}
+	if res.TB != nil {
+		t.Fatal("Expect returned TB is nil")
+	}
+}
+
+func TestKeyLookupNameNotFoundWithoutTB(t *testing.T) {
+	d := newDirectory(t, false)
+	_, err := d.Register(&RegistrationRequest{
+		Username: "alice",
+		Key:      []byte("key")})
+	if err != Success {
+		t.Fatal("Unable to register")
+	}
+	// lookup in the same epoch
+	// expect a proof of absence
+	res, err := d.KeyLookup(&KeyLookupRequest{Username: "alice"})
+	if err != ErrorNameNotFound ||
+		res.Error != ErrorNameNotFound {
+		t.Fatal("Expect error code", ErrorNameNotFound, "got", err)
+	}
+	if ap := res.AP; ap == nil || bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		t.Fatal("Expect a proof of absence")
+	}
+	if res.TB != nil {
+		t.Fatal("Expect returned TB is nil")
+	}
+
+	d.Update()
+	// lookup in different epochs
+	// expect a proof of absence
+	res, err = d.KeyLookup(&KeyLookupRequest{Username: "bob"})
+	if err != ErrorNameNotFound ||
+		res.Error != ErrorNameNotFound {
+		t.Fatal("Expect error code", ErrorNameNotFound, "got", err)
+	}
+	if ap := res.AP; ap == nil || bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		t.Fatal("Expect a proof of absence")
+	}
+	if res.TB != nil {
+		t.Fatal("Expect returned TB is nil")
+	}
+}
+
+func TestConsistencyCheck(t *testing.T) {
+	N := 10
+
+	d := newDirectory(t, false)
+	_, err := d.Register(&RegistrationRequest{
+		Username: "alice",
+		Key:      []byte("key")})
+	if err != Success {
+		t.Fatal("Unable to register")
+	}
+
+	d.Update()
+	savedSTR := d.LatestSTR().Signature
+	for i := 2; i < N; i++ {
+		d.Update()
+	}
+
+	// missed from epoch 2
+	res, err := d.ConsistencyCheck(&ConsistencyCheckRequest{"alice", uint64(2), uint64(11)})
```

Comment:
I'm a bit confused about the `uint64(11)` part. If `N = 10` at the beginning of the test and the update loop goes to `i < N` meaning that the loop ends when `N = 9`, how does and end epoch of 11 work? (The loop in `d.ConsistencyCheck()` retrieves all STRs up to the end epoch inclusive, but the latest epoch we have is 10).